binance-pipeline/lambdas/binance_historical_extraction.py

```diff
@@ -0,0 +1,249 @@
+    else:
+        candidate_id = trade_id + delta
+
+    return candidate_id
+
+
+def get_starting_ids(trade_id, concurrency, chunk_size, reverse=False):
+    starting_ids = []
+    for i in range(1, concurrency + 1):
+        delta = i*chunk_size
+        if (candidate_id := make_candidate_id(trade_id, delta, reverse)) >= 0:
+            starting_ids.append(candidate_id)
+
+    return starting_ids
+
+
+def handle_api_call(config: dict):
+    start = time.perf_counter()
+    payload = config['payload']
+
+    binance_query_api = BinanceQueryAPI()
+
+    status_code, res = binance_query_api.old_trade_lookup(
+                payload['symbol'], CHUNK_LIMIT, payload['fromId']
+            )
+
+    if status_code in [418, 429]:
+        handle_rate_limit(status_code, res, config)
+
+    # TO DO: store result to file
+    #
+    #
+
+    else:
+        delta = RESERVED_CONCURRENCY * CHUNK_LIMIT
+        next_trade_id = make_candidate_id(payload['fromId'], delta, config['reverse'])
+
+        if (len(res) == CHUNK_LIMIT) and (next_trade_id >= 0):
+            config['payload']['fromId'] = next_trade_id
+            message = {'config': config, 'type': 'api_call'}
+            end = time.perf_counter()
+
+            runtime = end - start
+            delay_seconds = calculate_delay_seconds(runtime, RESERVED_CONCURRENCY, RATE_LIMIT)
+
+            if delay_seconds:
+                delay_config = get_delay_config(delay_seconds)
+                message = {
+                            "status_code": status_code,
+                            "delay_config": delay_config,
+                            "config": config
+                        }
+                send_rate_throttle_message(message)
+            else:
+                send_sqs_message(SQS_QUEUE_NAME_EXTRACTION, message)
+
+    return
+
+
+def calculate_delay_seconds(runtime: int, reserved_concurrency: int, rate_limit: int) -> int:
+    delay_seconds = math.ceil(((60 * reserved_concurrency)/rate_limit) - runtime)
+    delay_seconds = max(0, delay_seconds)
+
+    return delay_seconds
+
+
+def handle_rate_limit(status_code: int, api_response: dict, config: dict):
+    """
+    when asked to wait x seconds before retrying API for throttling or ban, send to the delay queue
+    which will feed a separate delay lambda. The delay lambda will decrement the penalties and refeed
+    the delay queue until there are none left. When penalty time has expired, that delay lambda will
+    feed this lambda once again with the failed query config.
+    """
+    delay_config = get_delay_config(api_response['retry_after'])
+    message = {"status_code": status_code, "delay_config": delay_config, "config": config}
+    send_rate_throttle_message(message)
+
+    return
+
+
+def send_rate_throttle_message(message: dict):
+    delay_config = message['delay_config']
+    penalty_types = ['day', 'hour', 'minute', 'second']
+    sqs_queue_names = map(lambda x: f'SQS_QUEUE_NAME_DELAY_{x.upper()}', penalty_types)
```

Comment:
```python
penalty_queues = (delay_config[penalty], f'SQS_QUEUE_NAME_DELAY_{penalty.upper()}' for penalty in penalty_types)
```
Not sure if you can filter the generator with `if delay_config[penalty]`
that way the next loop basically takes the head element and calls `send_sqs_message` on it

Reply:
we can certainly do that - basically I only need to send to any one and only one queue for which there is at least one penalty. These queues feed a "delay lambda" (code not provided, trivial) which will read from the queue, decrement the payload for the delays by 1 for the queue it was fed by then feed back to any one and only one queue in the same way as the main lambda. It will do this until there are no penalties left to serve and when this is true feed back to the main lambda.

That's what the separate for and if block starting on 219 is for, I was hoping to make it explicit that I need to send to only one queue. If I smooth this into a more complex generator expression I think this crucial design information risks being lost - what are your thoughts?

Reply:
I only recommend the generator approach so that you avoid iterating twice, simplifying the code without reducing visibility.
Then you can do something like this:
```next((x for x in the_iterable if x > 3), default_value)```